### ISSUE TITLE
Generate rendering results for custom dataset

### ISSUE THREAD

**Issue by ZcsrenlongZ** [vye16/shape-of-motion#22]
Which command I should use to produce rendering videos for my custom data?

**Comment by doomzzju** [comment]
try run_rendering.py

**Comment by Bokai-Ji** [comment]
Firstly, run ./preproc/process_custom.py  to generate needed preprocessed elements (i.e., depth, reconstructions, ...)

Next, by running
```
python run_training.py \
    --work-dir <output dir> \ 
    --port <better not crash with the port in render, which is 8890 in default> \
    data:custom \
    --data.root-dir <the same data-root as in process_custom.py> \
    --data.seq-name <the same seq-name as in process_custom.py> \
```
, you will get a checkpoint file in `workdir/seq-name/checkpoints/`.

Finally, run
```
python run_rendering.py \
    --work-dir ./workdirs/seq-name/ \
    --port <optional> \
```
to see the result constructed from your custom data.

However, it will not render tracks in default. I checked the code and found the
```
import nvdiffrast.torch as dr
```
is commented and nvdiffrast is seemingly not installed. I'm yet trying to make this work.

**Comment by ZcsrenlongZ** [comment]
Many thanks!

**Comment by haoyi-duan** [comment]
Hi @Bokai-Ji, have you figured out how to render the tracks yet? Many thanks!

**Comment by Bokai-Ji** [comment]
> Hi @Bokai-Ji, have you figured out how to render the tracks yet? Many thanks!

Yes, I successfully rendered tracks by installing and uncommenting the import of `nvdiffrast`.

## Install `nvdiffrast`:

Clone the `nvdiffrast` repository by running the following command:
```
git clone https://github.com/NVlabs/nvdiffrast.git
```
You can clone it in any directory of your choice. Once cloned, navigate to the repository’s directory:
```
cd nvdiffrast
```
Install `nvdiffrast` as a PyPI package in your virtual environment by running:
```
pip install -e .
```
The installation process should complete without errors.

## Modify the Code:

In the file located at `/your-path/shape-of-motion/flow3d/vis/utils.py`, uncomment the following line:
```
import nvdiffrast.torch as dr
```

## Start the Rendering:

To begin rendering, run the following command:
```
python run_rendering.py \
    --work-dir ./workdirs/seq-name/ \
    --port <optional_port>
```
Ensure the Render tracks option is enabled to visualize the tracks.
<img width="1276" alt="tracks" src="https://github.com/user-attachments/assets/c0076b24-5723-498f-ab7f-094d3391f20b">

**Comment by haoyi-duan** [comment]
> > Hi @Bokai-Ji, have you figured out how to render the tracks yet? Many thanks!
> 
> Yes, I successfully rendered tracks by installing and uncommenting the import of `nvdiffrast`.
> 
> ## Install `nvdiffrast`:
> Clone the `nvdiffrast` repository by running the following command:
> 
> ```
> git clone https://github.com/NVlabs/nvdiffrast.git
> ```
> 
> You can clone it in any directory of your choice. Once cloned, navigate to the repository’s directory:
> 
> ```
> cd nvdiffrast
> ```
> 
> Install `nvdiffrast` as a PyPI package in your virtual environment by running:
> 
> ```
> pip install -e .
> ```
> 
> The installation process should complete without errors.
> 
> ## Modify the Code:
> In the file located at `/your-path/shape-of-motion/flow3d/vis/utils.py`, uncomment the following line:
> 
> ```
> import nvdiffrast.torch as dr
> ```
> 
> ## Start the Rendering:
> To begin rendering, run the following command:
> 
> ```
> python run_rendering.py \
>     --work-dir ./workdirs/seq-name/ \
>     --port <optional_port>
> ```
> 
> Ensure the Render tracks option is enabled to visualize the tracks. <img alt="tracks" width="1276" src="https://private-user-images.githubusercontent.com/56251146/356443132-c0076b24-5723-498f-ab7f-094d3391f20b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjMyNDg0ODIsIm5iZiI6MTcyMzI0ODE4MiwicGF0aCI6Ii81NjI1MTE0Ni8zNTY0NDMxMzItYzAwNzZiMjQtNTcyMy00OThmLWFiN2YtMDk0ZDMzOTFmMjBiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODEwVDAwMDMwMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc4YjIwYmJhZTU2NjA1M2Y5NDA3NTkzYzg0YmI5NzEzZjI1M2IzNmJiMGE4ZmJiYzhkYzYxNWM1NmE0MmVhNTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.raLdqbpLo26fzKDGfJkBi0ATdLj3jbXW-MWK--pqewA">

Thank you!

**Comment by enor2017** [comment]
Hi, thanks everyone for the discussion, and I can successfully view the tracks from a webpage. May I know how I can export a video similar to the iPhone dataset?

**Comment by Kairobo** [comment]
> **Step 1: Preprocess Custom Data (After mask obtained)**
> 
> First, generate the necessary preprocessed elements (such as depth and reconstructions) by running:
> 
> ```
> python process_custom.py --img-dirs ./data_root/images/seq-name/ --gpus 0
> ```
> 
> **Step 2: Train the Model**
> 
> Next, train the model by executing the following command:
> 
> ```
> python run_training.py \
>     --work-dir <output_dir> \
>     --port <port_number> \  # Ensure this port does not conflict with the render port, which defaults to 8890
>     data:custom \
>     --data.root-dir <data_root_dir> \
>     --data.seq-name <sequence_name>
> ```
> 
> This will create a checkpoint file in the `workdir/seq-name/checkpoints/` directory.
> 
> **Step 3: Render the Results**
> 
> Finally, render the results from your custom data by running:
> 
> ```
> python run_rendering.py \
>     --work-dir ./workdirs/seq-name/ \
>     --port <optional_port>
> ```
> 
> **Note:** By default, the tracks will not be rendered. Upon inspecting the code, it appears that the line:
> 
> ```
> import nvdiffrast.torch as dr
> ```
> 
> is commented out, and the `nvdiffrast` package does not seem to be installed. Efforts are ongoing to resolve this issue.

It seems the run_training.py is modified so that it cannot run the second step again, any help on that?

**Comment by Kairobo** [comment]
> > **Step 1: Preprocess Custom Data (After mask obtained)**
> > First, generate the necessary preprocessed elements (such as depth and reconstructions) by running:
> > ```
> > python process_custom.py --img-dirs ./data_root/images/seq-name/ --gpus 0
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > **Step 2: Train the Model**
> > Next, train the model by executing the following command:
> > ```
> > python run_training.py \
> >     --work-dir <output_dir> \
> >     --port <port_number> \  # Ensure this port does not conflict with the render port, which defaults to 8890
> >     data:custom \
> >     --data.root-dir <data_root_dir> \
> >     --data.seq-name <sequence_name>
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This will create a checkpoint file in the `workdir/seq-name/checkpoints/` directory.
> > **Step 3: Render the Results**
> > Finally, render the results from your custom data by running:
> > ```
> > python run_rendering.py \
> >     --work-dir ./workdirs/seq-name/ \
> >     --port <optional_port>
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > **Note:** By default, the tracks will not be rendered. Upon inspecting the code, it appears that the line:
> > ```
> > import nvdiffrast.torch as dr
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > is commented out, and the `nvdiffrast` package does not seem to be installed. Efforts are ongoing to resolve this issue.
> 
> It seems the run_training.py is modified so that it cannot run the second step again, any help on that?

I solved it. we can actually change the causualdata.py for it
